Move analyze phase before detect (#135)
buildpacks/rfcs: text/0000-prepare-phase.md

- RFC Pull Request: (leave blank)
- CNB Pull Request: [buildpacks/spec#176](https://github.com/buildpacks/spec/pull/176)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

This is a proposal for a new Lifecycle phase, called "prepare", that would run before all other phases.

# Definitions
[definitions]: #definitions

* __stack descriptor__ - an enhanced version of the [`stack.toml`](https://github.com/buildpacks/spec/blob/main/platform.md#stacktoml-toml)
* __project descriptor__ - the [`project.toml`](https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md) extension specification

# Motivation
[motivation]: #motivation

The prepare phase would support the following features and capabilities:
* [Stack buildpacks](https://github.com/buildpacks/rfcs/pull/111), which require a phase to read run-image mixins validation prior to detection
* [Inline buildpacks](https://github.com/buildpacks/rfcs/blob/main/text/0048-inline-buildpack.md), which require parsing of the `project.toml` in the lifecycle
* [Lifecycle configuration](https://github.com/buildpacks/rfcs/pull/128)

# What it is
[what-it-is]: #what-it-is

> The prepare phase will run before all other phase, and prepare the execution environment for a buildpack build. This phase will have access to secrets and credentials used to access registries and other services.

Reviewer: ```suggestion
The prepare phase will run before all other phases, and prepare the execution environment for a buildpack build. This phase will have access to secrets and credentials used to access registries and other services.
```

> The prepare phase will run before all other phase, and prepare the execution environment for a buildpack build. This phase will have access to secrets and credentials used to access registries and other services.

Reviewer: > "access to secrets and credentials used to access registries and other services"
Is this something that will happen throughout the lifecycle or just in prepare?

Author: it happens in analyze, restore, and export (but not detect and build)

The [logic in `pack` that parses a `project.toml`](https://github.com/buildpacks/pack/blob/main/project/project.go) would be copied or moved into the `preparer`.

The app source code (which may be provided to the prepare either as a directory, volume, or tarball) would be mutated (either by copying it to a new location, or making changes directory against it). The `preparer` may delete files to apply the include and exclude rules from `project.toml`.

# Drawbacks
[drawbacks]: #drawbacks

* Yet another lifecycle phase

# Alternatives
[alternatives]: #alternatives

- [Reverse the order of analyze and detect phases](https://github.com/buildpacks/spec/pull/172)

# Prior Art
[prior-art]: #prior-art

- [Tekton prepare step](https://github.com/tektoncd/catalog/blob/11a17cfe87779099b0b61be3f1e496dfa79646b3/task/buildpacks-phases/0.1/buildpacks-phases.yaml#L61-L78)

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

- Do we still need `analyzer`, or can the remaining parts of analyze phase be rolled into restore phase?
- Does `pack` still need to parse `project.toml`, or is there value in reading it early on (before lifecycle runs)?

> - Should we create a shared library for `project.toml` parsing?

Reviewer: Out of scope of the RFC, but this sounds great.

# Definitions
[definitions]: #definitions

* __stack descriptor__ - an enhanced version of the [`stack.toml`](https://github.com/buildpacks/spec/blob/main/platform.md#stacktoml-toml)
* __project descriptor__ - the [`project.toml`](https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md) extension specification

# Motivation
[motivation]: #motivation

The prepare phase would support the following features and capabilities:
* [Stack buildpacks](https://github.com/buildpacks/rfcs/pull/111), which require a phase to read run-image mixins validation prior to detection
* [Inline buildpacks](https://github.com/buildpacks/rfcs/blob/main/text/0048-inline-buildpack.md), which require parsing of the `project.toml` in the lifecycle
* [Lifecycle configuration](https://github.com/buildpacks/rfcs/pull/128)

# What it is
[what-it-is]: #what-it-is

The prepare phase will run before all other phases, and prepare the execution environment for a buildpack build. This phase will have access to secrets and credentials used to access registries and other services.

## Responsibilities

* Stack validation, to ensure that a new run-image is campatible with the previous app image

> * Retrive run-image mixins, which will be use dby subsequent phases

Reviewer 2: ```suggestion
* Retrieve identifier (imageID or digest), stack ID, and mixins, which will be used by subsequent phases
```
We should retrieve all relevant data. Including an immutable ID that can be used by subsequent phases will help us avoid race conditions in the case where a new run image is pushed to the same tag between build phases.

* Validation of registry credentials, to avoid a long build that fails during export phase
* Parsing the project descriptor and performance various operations based on its contents, include:
    - downloading buildpacks
    - creating ephemeral buildpacks
    - applying include and exclude rules

>     - modifying environment variables

Reviewer 2: ```suggestion
    - adding environment variables to <platform>/env
```

## Inputs

* Log level
* Run-image
* Stack ID
* Project descriptor (optional)

> * App source code

Reviewer 2: ```suggestion
* App source code
* Previous Image
* Destination tag
```
To preform the responsibilities above will will need these additional inputs.

## Output

* Exit status
* Info-level logs to `stdout`
* Error-level logs to `stderr`

> * Stack descriptor

Reviewer 2: I don't think we should mutate the `stack.toml` that is included on the builder, but should instead include run image info in the analysis metadata (or a new file).

* Analysis metadata
* Buildpacks (derived from inline buildpacks in project descriptor, or buildpacks in project descriptor that are not present in the builder)
* Order definition (derived from buildpacks configuration in project descriptor)
* Lifecycle configuration (derived from configuration in project descriptor)

> * Mutated app source code (applying include and exclude rules in project descriptor)

Reviewer 2: Should we output [project metadata](https://github.com/buildpacks/spec/blob/main/platform.md#project-metadatatoml-toml)?
```suggestion
* Mutated app source code (applying include and exclude rules in project descriptor)
* Project metadata 
```
I imagine this would eventually be enhanced with other fields like `authors` and `documentaion-url` from `project.toml`. We could probably skip it for now but eventually we may want to add source metadata detection logic into the lifecycle as well.

Reviewer 3: Bumped to future work (all of project.toml)

# How it Works
[how-it-works]: #how-it-works

The prepare phase would be implemented as the `/cnb/lifecycle/preparer` binary. A platform MUST execute this phases either by invoking the `/cnb/lifecycle/preparer` binary or by executing `/cnb/lifecycle/creator`.

The `preparer` binary will have access to the [`Keychain`](https://github.com/buildpacks/lifecycle/blob/main/auth/env_keychain.go), and MUST NOT execute arbitrary code provided by either the buildpack user or buildpack author.

The [logic in the `analyzer` phase that reads image metadata and outputs an `analyzed.toml`](https://github.com/buildpacks/lifecycle/blob/main/analyzer.go#L34-L40) would be moved into the `preparer`.

The [logic in `pack` that parses a `project.toml`](https://github.com/buildpacks/pack/blob/main/project/project.go) would be copied or moved into the `preparer`.

> The app source code (which may be provided to the prepare either as a directory, volume, or tarball) would be mutated (either by copying it to a new location, or making changes directory against it). The `preparer` may delete files to apply the include and exclude rules from `project.toml`.

Reviewer 2: ```suggestion
The app source code (which may be provided to the prepare either as a directory, volume, or tarball) would be mutated (either by copying it to a new location, or making changes directly). The `preparer` may delete files to apply the include and exclude rules from `project.toml`.
```

# Drawbacks
[drawbacks]: #drawbacks

* Yet another lifecycle phase

# Alternatives
[alternatives]: #alternatives

- [Reverse the order of analyze and detect phases](https://github.com/buildpacks/spec/pull/172)

> - Split analyze into two parts (one to write `analyzed.toml` and one to analyze layers). The first (`analyzed.toml`) would be moved before the detect phase, and the second part (analyzing layers) would be merged into the restore phase.

Reviewer 2: Just to clarify, the differences between the proposal and this alternative are the following?
1. first phase is named `preparer` instead of `analyzer`
1. `analyzer` still exists as a required phase distinct from `restorer`

# Prior Art
[prior-art]: #prior-art

- [Tekton prepare step](https://github.com/tektoncd/catalog/blob/11a17cfe87779099b0b61be3f1e496dfa79646b3/task/buildpacks-phases/0.1/buildpacks-phases.yaml#L61-L78)

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

> - Do we still need `analyzer`, or can the remaining parts of analyze phase be rolled into restore phase?

Reviewer 2: It's worth calling out that even if we merge this functionality into `restorer` in the new API, if we name the new phase `restorer`, we will still need to ship an `anlayzer` binary to maintain support for older platform APIs.
